sdk/nodejs/resource.ts

```diff
@@ -676,3 +676,160 @@ export class ComponentResource extends Resource {
+ *  4. For the purposes of merging `dependsOn`, `provider` and `providers` are always treated as
+ *     collections, even if only a single value was provided.
+ */
+export function mergeOptions(opts1: CustomResourceOptions | undefined, opts2: CustomResourceOptions | undefined): CustomResourceOptions;
+export function mergeOptions(opts1: ComponentResourceOptions | undefined, opts2: ComponentResourceOptions | undefined): ComponentResourceOptions;
+export function mergeOptions(opts1: ResourceOptions | undefined, opts2: ResourceOptions | undefined): ResourceOptions;
+export function mergeOptions(opts1: ResourceOptions | undefined, opts2: ResourceOptions | undefined): ResourceOptions {
+    const dest = <any>{ ...opts1 };
+    const source = <any>{ ...opts2 };
+
+    // Ensure provider/providers are all expanded into the `{ provName: prov }` form.
+    // This makes merging simple.
+    expandProviders(dest);
+    expandProviders(source);
+
+    // iterate specifically over the supplied properties in [source].  Note: there may not be an
+    // corresponding value in [dest].
+    for (const key of Object.keys(source)) {
+        const destVal = dest[key];
+        const sourceVal = source[key];
+
+        if (key === "providers") {
+            // Note: this expansion is safe to do as expandProviders will have made sure
+            // that both collections are in Record<string, ProviderResource> form.
+            dest[key] = { ...destVal, ...sourceVal };
+            continue;
+        }
+
+        // Due to the possibility of top level and nested Inputs for 'dependsOn' we have to handle
+        // that property specially.
+        if (key === "dependsOn") {
+            dest[key] = mergeDependsOn(destVal, sourceVal);
+            continue;
+        }
+
+        // we should have no promises/outputs at top level for any other resource option properties.
+        if (isPromiseOrOutput(destVal)) {
+            throw new Error(`Unexpected promise/output in opts1.${key}`);
+        }
+
+        if (isPromiseOrOutput(sourceVal)) {
+            throw new Error(`Unexpected promise/output in opts2.${key}`);
+        }
+
+        dest[key] = merge(destVal, sourceVal);
+    }
+
+    // Now, if we are left with a .providers that is just a single key/value pair, then
+    // collapse that down into .provider form.
+    collapseProviders(dest);
+
+    return dest;
+}
+
+function isPromiseOrOutput(val: any): boolean {
+    return val instanceof Promise || Output.isInstance(val);
+}
+
+function merge(dest: any, source: any): any {
+    // if the second options bag contained `prop: null` or `prop: undefined` then that overrides
+    // anything in the destination.
+
+    if (source === null || source === undefined) {
+        return source;
+    }
+
+    // if there's no destination value, the source value wins.
+    if (dest === null || dest === undefined) {
+        return source;
+    }
```

Comment:
Do these checks really need to be separate?  Can’t mergeArrays just handle undefed|null inputs?

Reply:
you're right. t his was from a more complicated starting point.  it can be much simplified now.

Reply:
Done.  Let me know what you think of the current form.